python/cudf/cudf/core/buffer.py

```diff
@@ -5,108 +5,206 @@
-            self._ptr = data.ptr
-            self._size = data.size
-            self._owner = data
-        elif hasattr(data, "__array_interface__") or hasattr(
-            data, "__cuda_array_interface__"
-        ):
-            self._init_from_array_like(data, owner)
-        elif isinstance(data, memoryview):
-            self._init_from_array_like(np.asarray(data), owner)
-        elif isinstance(data, int):
-            if not isinstance(size, int):
-                raise TypeError("size must be integer")
+        if isinstance(data, int):
+            if size is None:
+                raise ValueError(
+                    "size must be specified when `data` is an integer"
+                )
             self._ptr = data
             self._size = size
             self._owner = owner
-        elif data is None:
-            self._ptr = 0
-            self._size = 0
-            self._owner = None
         else:
-            try:
-                data = memoryview(data)
-            except TypeError:
-                raise TypeError("data must be Buffer, array-like or integer")
-            self._init_from_array_like(np.asarray(data), owner)
-
-    @classmethod
-    def from_buffer(cls, buffer: Buffer, size: int = None, offset: int = 0):
-        """
-        Create a buffer from another buffer
-
-        Parameters
-        ----------
-        buffer : Buffer
-            The base buffer, which will also be set as the owner of
-            the memory allocation.
-        size : int, optional
-            Size of the memory allocation (default: `buffer.size`).
-        offset : int, optional
-            Start offset relative to `buffer.ptr`.
-        """
+            if size is not None or owner is not None:
+                raise ValueError(
+                    "`size` and `owner` must be None when "
+                    "`data` is an buffer-like object"
+                )
+
+            # `data` is a buffer-like object
+            buf: Any = data
+            if isinstance(buf, rmm.DeviceBuffer):
+                self._ptr, self._size, self._owner = buf.ptr, buf.size, buf
+                return
+            iface = getattr(buf, "__cuda_array_interface__", None)
+            if iface:
+                ptr, size = get_ptr_and_size(iface)
+                self._ptr, self._size, self._owner = ptr, size, buf
+                return
+            ptr, size = get_ptr_and_size(np.asarray(buf).__array_interface__)
+            buf = rmm.DeviceBuffer(ptr=ptr, size=size)
+            self._ptr, self._size, self._owner = buf.ptr, buf.size, buf
+
+    def __getitem__(self, key: Union[int, slice]) -> Buffer:
+        if isinstance(key, int):
+            if key >= self.size:
+                raise TypeError("index out of bounds")
+            key = slice(key, key + 1)
+        if not isinstance(key, slice):
+            raise TypeError("index must be an int or a slice")
+        start, stop, step = key.indices(self.size)
```

Comment:
This call to `key.indices(self.size)` will wrap negative indices to positive values, but we probably still need to validate `stop >= start` somewhere. (Does the Buffer constructor need to validate `size >= 0` instead of doing it here?)

Reply:
Done in the Buffer constructor